src/uct/tcp/tcp_listener.c

```diff
@@ -154,7 +154,32 @@ UCS_CLASS_CLEANUP_FUNC(uct_tcp_listener_t)
 ucs_status_t uct_tcp_listener_reject(uct_listener_h listener,
                                      uct_conn_request_h conn_request)
 {
-    return UCS_ERR_NOT_IMPLEMENTED;
+    uct_tcp_sockcm_ep_t *cep         = (uct_tcp_sockcm_ep_t *)conn_request;
+    uct_tcp_sockcm_t *tcp_sockcm     = uct_tcp_sockcm_ep_get_cm(cep);
+    char peer_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+
+    UCS_ASYNC_BLOCK(tcp_sockcm->super.iface.worker->async);
+
+    ucs_assert((cep->state & UCT_TCP_SOCKCM_EP_ON_SERVER) &&
+               !(cep->state & (UCT_TCP_SOCKCM_EP_SERVER_CREATED |
+                               UCT_TCP_SOCKCM_EP_FAILED)));
```

Comment:
why not EP_FAILED? it can fail before we reject it

Reply:
but if there is a failure, we remove the fd from the handlers list and return without calling cep->listener->conn_request_cb

Reply:
do you mean that after we called cep->listener->conn_request_cb, we will no longer get errors on it?
so how we can have UCT_TCP_SOCKCM_EP_GOT_DISCONNECT flag which is checked later?

Reply:
1. yes.
2. if uct_tcp_listener_reject was called from the main thread but a disconnect event arrived on the async thread.

Reply:
(2) seems like a contradiction of (1): if we no longer get events, how could we get disconnect?

Reply:
what would prevent the user from calling uct_listenet_reject from the main thread after getting a disconnect from the client on the async thread?

Reply:
seems the behavior "if there is a failure, we remove the fd from the handlers list and return without calling cep->listener->conn_request_cb" is incorrect
in such case, we need to call cep->listener->conn_request_cb with error status

Reply:
currently conn_request_cb doesnt have status in its args - https://github.com/openucx/ucx/blob/master/src/uct/api/uct_def.h#L294
didn't we say that in case of an error on the server before it created an ep with uct_ep_create, we drop the connection request?

Reply:
> didn't we say that in case of an error on the server before it created an ep with uct_ep_create, we drop the connection request?

this is right. so doesn't it mean that we also can't have UCT_TCP_SOCKCM_EP_GOT_DISCONNECT flag on on the server side?

Reply:
what do you mean by 'can't have'? 
wording of GOT_DISCONNECT on a not yet created ep or how can this flag already be set in the reject function?
if it's the second case, then since there is nothing that can prevent the server from calling uct_listenet_reject from the main thread after getting a disconnect from the client on the async thread, i added this check, so that the server doesnt reject if it already received a disconnect event. makes sense?

Reply:
IMO we should unite the states GOT_DISCONNECT and FAILED. They mean the same thing: that remote socket was closed/connection was broken.

Reply:
but FAILED is for errors that are not only related to the connection with the peer, it's for local errors (like for example, an error from ucs_sockaddr_get_ifname()). in this case, there is no disconnect notice from the peer so it's not really the same...
anyway, uniting these flags isn't related to this PR, can it be further discussed after it?

Reply:
@yosefe wdyt?